esmvaltool/utils/recipe_filler.py

```diff
@@ -0,0 +1,684 @@
+
+            # Globs all the wildcards into a list of files.
+            files = glob(new_path)
+            all_files.extend(files)
+    if not all_files:
+        logger.warning("Could not find any file for data specifications.")
+
+    return all_files
+
+
+def _file_to_recipe_dataset(fn, cmip_era, file_dict):
+    """Convert a filename to an recipe ready dataset."""
+    # Add the obvious ones - ie the one you requested!
+    output_dataset = {}
+    output_dataset['project'] = cmip_era
+    for key, value in file_dict.items():
+        if value == '*':
+            continue
+        if key in dataset_order:
+            output_dataset[key] = value
+
+    # Split file name and base path into directory structure and filenames.
+    basefiles = _determine_basepath(cmip_era)
+    _, fnfile = os.path.split(fn)
+
+    for basefile in basefiles:
+        _, basefile = os.path.split(basefile)
+        # Some of the key words include the splitting character '_' !
+        basefile = basefile.replace('short_name', 'shortname')
+        basefile = basefile.replace('start_year', 'startyear')
+        basefile = basefile.replace('end_year', 'endyear')
+
+        # Assume filename is separated by '_'
+        basefile_split = [key.replace("{", "") for key in basefile.split('_')]
+        basefile_split = [key.replace("}", "") for key in basefile_split]
+        fnfile_split = fnfile.split('_')
+
+        # iterate through directory structure looking for useful bits.
+        for base_key, fn_key in zip(basefile_split, fnfile_split):
+            if base_key == '*.nc':
+                fn_key = fn_key.replace('.nc', '')
+                start_year, end_year = fn_key.split('-')
+                output_dataset['start_year'] = start_year
+                output_dataset['end_year'] = end_year
+            elif base_key == "ensemble*.nc":
+                output_dataset['ensemble'] = fn_key
+            elif base_key == "grid*.nc":
+                output_dataset['grid'] = fn_key
+            elif base_key not in ["shortname", "ensemble*.nc", "*.nc"]:
+                output_dataset[base_key] = fn_key
+    if "exp" in file_dict:
+        if isinstance(file_dict["exp"], list):
+            output_dataset["exp"] = file_dict["exp"]
+
+    return output_dataset
+
+
+def _remove_duplicates(add_datasets):
+    """
+    Remove accidental duplicates.
+
+    Close to 0% chances this will ever be used.
+    May be used when there are actual duplicates in data
+    storage, we've seen these before, but seldom.
+    """
+    datasets = []
+    seen = set()
+
+    for dataset in add_datasets:
+        orig_exp = dataset["exp"]
+        dataset["exp"] = str(dataset["exp"])
+        tup_dat = tuple(dataset.items())
+        if tup_dat not in seen:
+            seen.add(tup_dat)
+            dataset["exp"] = orig_exp
+            datasets.append(dataset)
+
+    return datasets
+
+
+def _check_recipe(recipe_dict):
```

Comment:
We have already have a recipe checker in `esmvalcore`: https://github.com/ESMValGroup/ESMValCore/blob/master/esmvalcore/_recipe_checks.py

Why not use that?

Reply:
see my comment above :+1: